lib/Dialect/ESI/runtime/cpp/lib/backends/Trace.cpp

```diff
@@ -271,12 +275,63 @@ TraceAccelerator::requestChannelsFor(AppIDPath idPath,
   return impl->requestChannelsFor(idPath, bundleType);
 }
 
+class TraceHostMem : public HostMem {
+public:
+  TraceHostMem(TraceAccelerator::Impl &impl) : impl(impl) {}
+
+  struct TraceHostMemRegion : public HostMemRegion {
+    TraceHostMemRegion(std::size_t size, TraceAccelerator::Impl &impl)
+        : impl(impl) {
+      ptr = malloc(size);
+      this->size = size;
+    }
+    virtual ~TraceHostMemRegion() {
+      impl.write("HostMem") << "free " << ptr << std::endl;
+      free(ptr);
+    }
+    virtual void *getPtr() const { return ptr; }
+    virtual std::size_t getSize() const { return size; }
+
+  private:
+    void *ptr;
+    std::size_t size;
+    TraceAccelerator::Impl &impl;
+  };
+
+  virtual std::unique_ptr<HostMemRegion> allocate(std::size_t size,
+                                                  HostMem::Options opts) const {
+    auto ret =
+        std::unique_ptr<HostMemRegion>(new TraceHostMemRegion(size, impl));
+    impl.write("HostMem") << ret->getPtr() << " allocate " << size
```

Comment:
Nit: a void* is printed to a stream as hex, right? If so (and i'd prefer that), suggestion to prefix it with `0x` (i don't assume the stream itself infers the prefix).